runtime/src/mm/cpp/ObjectFactory.hpp

```diff
@@ -0,0 +1,280 @@
+/*
+ * Copyright 2010-2020 JetBrains s.r.o. Use of this source code is governed by the Apache 2.0 license
+ * that can be found in the LICENSE file.
+ */
+
+#ifndef RUNTIME_MM_OBJECT_FACTORY_H
+#define RUNTIME_MM_OBJECT_FACTORY_H
+
+#include <mutex>
+
+#include "Alignment.hpp"
+#include "Alloc.h"
+#include "Memory.h"
+#include "Mutex.hpp"
+#include "Utils.hpp"
+
+namespace kotlin {
+namespace mm {
+
+namespace internal {
+
+// A queue that is constructed by collecting subqueues from several `Producer`s.
+// This is essentially a heterogeneous `MultiSourceQueue` on top of a singly linked list that
+// uses `konanAllocMemory` and `konanFreeMemory`
+// TODO: Consider merging with `MultiSourceQueue` somehow.
+class ObjectFactoryStorage : private Pinned {
+public:
+    // This class does not know its size at compile-time.
+    class Node : private Pinned {
+    public:
+        ~Node() = default;
+
+        static void operator delete(void* ptr) noexcept { konanFreeMemory(ptr); }
+
+        // Note: This can only be trivially destructible data, as nobody can invoke its destructor.
+        void* Data() noexcept { return this + 1; }
```

Comment:
This function doesn't align the result. What is `dataAlignment` then?

Reply:
Right, reworked aligning. Now it should be more clear and correct.